### PR TITLE
[FIXED JENKINS-41246] Guard against PRs from deleted forks

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-41246](https://issues.jenkins-ci.org/browse/JENKINS-41246)

@reviewbybees

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/github_branch_source/PullRequestSCMHead.java [hunk 1]
```diff
@@ -61,8 +62,9 @@
         this.number = pr.getNumber();
         this.target = new BranchSCMHead(pr.getBase().getRef());
         // the source stuff is immutable for a pull request on github, so safe to store here
-        this.sourceOwner = pr.getHead().getRepository().getOwnerName();
-        this.sourceRepo = pr.getHead().getRepository().getName();
+        GHRepository repository = pr.getHead().getRepository(); // may be null for deleted forks JENKINS-41246
+        this.sourceOwner = repository.getOwnerName();
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/PullRequestSCMHead.java]: So, did you mean to check for null??

Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/PullRequestSCMHead.java]: Trying to ascertain reasons about code I wrote more than 15 minutes ago is an exercise in futility... I suspect I did mean to check for `null` though... good catch

File: src/main/java/org/jenkinsci/plugins/github_branch_source/PullRequestSCMHead.java [hunk 2]
```diff
@@ -61,8 +62,9 @@
         this.number = pr.getNumber();
         this.target = new BranchSCMHead(pr.getBase().getRef());
         // the source stuff is immutable for a pull request on github, so safe to store here
-        this.sourceOwner = pr.getHead().getRepository().getOwnerName();
-        this.sourceRepo = pr.getHead().getRepository().getName();
+        GHRepository repository = pr.getHead().getRepository(); // may be null for deleted forks JENKINS-41246
+        this.sourceOwner = repository == null ? null : repository.getOwnerName();
+        this.sourceRepo = repository == null ? null : repository.getName();
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/PullRequestSCMHead.java]: Suggest marking getters `@CheckForNull`.

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java [hunk 3]
```diff
@@ -619,12 +619,14 @@ private void doRetrieve(SCMSourceCriteria criteria, SCMHeadObserver observer, Ta
...
-                        && collaboratorNames.contains(ghPullRequest.getHead().getRepository().getOwnerName());
+                GHRepository repository = ghPullRequest.getHead().getRepository();
+                // repository may be null for deleted forks (JENKINS-41246) in which case they are not trusted
+                boolean trusted = repository != null && collaboratorNames != null
+                        && collaboratorNames.contains(repository.getOwnerName());
                 if (!trusted) {
                     listener.getLogger().format("    (not from a trusted source)%n");
                 }
-                for (boolean merge : new boolean[] {false, true}) {
+                for (boolean merge : new boolean[]{false, true}) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java]: gratuitous

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java [hunk 4]
```diff
@@ -671,7 +673,7 @@ private void doRetrieve(SCMSourceCriteria criteria, SCMHeadObserver observer, Ta
                             user.getLogin(),
                             user.getName(),
                             user.getEmail()
-                            ));
+                    ));
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java]: gratuitous

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java [hunk 5]
```diff
@@ -708,10 +710,12 @@ private void doRetrieve(SCMSourceCriteria criteria, SCMHeadObserver observer, Ta
                     } else {
                         baseHash = ghPullRequest.getBase().getSha();
                     }
-                    PullRequestSCMRevision rev = new PullRequestSCMRevision(head, baseHash, ghPullRequest.getHead().getSha());
+                    PullRequestSCMRevision rev =
+                            new PullRequestSCMRevision(head, baseHash, ghPullRequest.getHead().getSha());
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java]: gratuitous

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java [hunk 6]
```diff
@@ -708,10 +710,12 @@ private void doRetrieve(SCMSourceCriteria criteria, SCMHeadObserver observer, Ta
...
                     observer.observe(head, rev);
                     if (!observer.isObserving()) {
-                        listener.getLogger().format("%n  %d pull requests were processed (query completed)%n", pullrequests);
+                        listener.getLogger()
+                                .format("%n  %d pull requests were processed (query completed)%n", pullrequests);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java]: gratuitous

**Pull request merged**